File: Samples/Objective-C/SignatureExample/SignatureExample/FileBrowser.m
//
//  FileBrowser.m
//  SignatureExample
//
//  Created by Joss Giffard-Burley on 06/08/2014.
//  Copyright (c) 2014 Wacom. All rights reserved.
//

#import "FileBrowser.h"
#import "UIImage+TintImage.h"

@interface FileBrowser ()

@property (strong, nonatomic) IBOutlet UIButton *cancelButton;
@property (strong, nonatomic) IBOutlet UIButton *okButton;
@property (strong, nonatomic) UIColor *tintColour;
@property (strong, nonatomic) IBOutlet UITableView *tableView;
@property (strong, nonatomic) IBOutlet UILabel *titleLabel;
@property (strong, nonatomic) IBOutlet UITextField *inputTextField;
@property (strong, nonatomic) IBOutlet UIView *textFieldUnderline;
@property (strong, nonatomic) IBOutlet UIImageView *logo;
@property (strong, nonatomic) NSString *currentDirectory;
@property (strong, nonatomic) NSArray *fileList;
@property (strong, nonatomic) NSArray *signatureExtensions;
@property (strong, nonatomic) UIImage *directoryIcon;
@property (strong, nonatomic) UIImage *textIcon;
@property (strong, nonatomic) UIImage *imageIcon;
@property (strong, nonatomic) UIImage *binaryIcon;


- (IBAction)dismissKeyboard;

- (void)updateFileList;

@end

@implementation FileBrowser


- (void)viewDidLoad {
    [super viewDidLoad];
    self.okButton.layer.cornerRadius = 5.0f;
    self.cancelButton.layer.cornerRadius = 5.0f;
    self.tintColour = [UIColor colorWithRed:0.0f green:0.592 blue:0.831 alpha:1.0f];
    self.tableView.layer.cornerRadius = 3.0f;
    self.tableView.layer.borderColor = [UIColor colorWithRed:0.7f green:0.7f blue:0.7f alpha:1.0f].CGColor;
    self.tableView.layer.borderWidth = 1.0f;
    //Tint the logo
    [self.logo setImage:[[UIImage imageNamed:@"logo_window"] tintedImageWithColour:[UIColor colorWithRed:0.4f
                                                                                                        green:0.4f
                                                                                                         blue:0.4f
                                                                                                        alpha:1.0f]]];
    self.currentDirectory = self.baseDirectory;
    self.signatureExtensions = @[@"png", @"txt", @"fss"];
    
    //Create icons for list
    self.directoryIcon = [[UIImage imageNamed:@"action_collection"] tintedImageWithColour:[UIColor colorWithRed:0.121f green:0.494f blue:0.807f alpha:1.0f]];
    self.textIcon = [[UIImage imageNamed:@"text_file"] tintedImageWithColour:[UIColor colorWithRed:0.121f green:0.494f blue:0.807f alpha:1.0f]];
    self.imageIcon = [[UIImage imageNamed:@"action_picture"] tintedImageWithColour:[UIColor colorWithRed:0.121f green:0.494f blue:0.807f alpha:1.0f]];
    self.binaryIcon = [[UIImage imageNamed:@"binary_file"] tintedImageWithColour:[UIColor colorWithRed:0.121f green:0.494f blue:0.807f alpha:1.0f]];

    
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    self.titleLabel.text = self.titleText;
    self.inputTextField.text = self.selectedFilename;
    
    [self updateFileList];
}

- (IBAction)dismissKeyboard {
    [self.inputTextField resignFirstResponder];
}

#pragma mark - Interface methods

- (void)setSignatureTypesOnly:(BOOL)signatureTypesOnly {
    _signatureTypesOnly = signatureTypesOnly;
    [self updateFileList];
}

- (void)updateFileList {
    NSFileManager *fm = [NSFileManager defaultManager];
    NSDirectoryEnumerator *dirEnum = [fm enumeratorAtPath:self.currentDirectory];
    NSMutableArray *files = [[NSMutableArray alloc] init];
    NSPredicate *extensionPredicate = [NSPredicate predicateWithFormat:@"SELF IN %@", self.signatureExtensions];
    
    NSString *currentFile;
    while(currentFile = [dirEnum nextObject]) {
        
        //Skip files starting '.' and subdirs
        if([currentFile characterAtIndex:0] == '.' ||  [currentFile rangeOfString:@"/"].length > 0) {
            continue;
        }
        
        if(self.signatureTypesOnly) {
            if(dirEnum.fileAttributes[NSFileType] == NSFileTypeDirectory || [extensionPredicate evaluateWithObject:[currentFile pathExtension]]) {
                NSMutableDictionary *file = [NSMutableDictionary dictionaryWithDictionary:dirEnum.fileAttributes];
                [file setObject:[currentFile pathExtension] forKey:@"PATH_EXT"];
                [file setObject:[currentFile lastPathComponent] forKey:@"FILENAME"];
                [files addObject:file];
            }
        } else {
            NSMutableDictionary *file = [NSMutableDictionary dictionaryWithDictionary:dirEnum.fileAttributes];
            [file setObject:[currentFile pathExtension] forKey:@"PATH_EXT"];
            [file setObject:[currentFile lastPathComponent] forKey:@"FILENAME"];
            [files addObject:file];
        }
    }
    //Sort file list A->Z
    
    [files sortUsingComparator:^NSComparisonResult(NSDictionary *obj1, NSDictionary *obj2) {
        return([obj1[@"FILENAME"] compare:obj2[@"FILENAME"] options:NSCaseInsensitiveSearch]);
    }];
    
    self.fileList = [NSArray arrayWithArray:files];
    [self.tableView reloadData];
}

#pragma mark - UITableView methods


- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if([self.baseDirectory isEqualToString:self.currentDirectory]) {
        return(self.fileList.count);
    } else {
        return(self.fileList.count + 1);
    }
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return(1);
}


- (UIView *)tableView:(UITableView *)tableView viewForFooterInSection:(NSInteger)section {
    return([UIView new]);
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"FILEBROWSER" forIndexPath:indexPath];
    NSInteger idx = [self.baseDirectory isEqualToString:self.currentDirectory] ? indexPath.row : indexPath.row - 1; //Add a ".." entry for sub dirs at the top of the list
    NSPredicate *imageExts = [NSPredicate predicateWithFormat:@"SELF IN %@", @[@"jpg", @"png", @"bmp", @"jpeg", @"tiff", @"gif"]];
    NSPredicate *docExts = [NSPredicate predicateWithFormat:@"SELF IN %@", @[@"txt", @"doc", @"pdf", @"docx", @"pages"]];
    
    
    if(indexPath.row == 0 && ![self.baseDirectory isEqualToString:self.currentDirectory]) {
        cell.textLabel.text = @"..";
        cell.accessoryType = UITableViewCellAccessoryNone;
        cell.imageView.image = self.directoryIcon;
    } else {
        NSDictionary *fileDetails = self.fileList[idx];
        
        if(fileDetails[NSFileType] == NSFileTypeDirectory) { //Directory
            cell.imageView.image = self.directoryIcon;
            cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
        } else if([imageExts evaluateWithObject:fileDetails[@"PATH_EXT"]]) { //Image file
            cell.imageView.image = self.imageIcon;
            cell.accessoryType = UITableViewCellAccessoryNone;
        } else if([docExts evaluateWithObject:fileDetails[@"PATH_EXT"]]) { //Txt / doc file
            cell.imageView.image = self.textIcon;
            cell.accessoryType = UITableViewCellAccessoryNone;
        } else { //Give it the binary icon
            cell.accessoryType = UITableViewCellAccessoryNone;
            cell.imageView.image = self.binaryIcon;
        }
        
        cell.textLabel.text = fileDetails[@"FILENAME"];
    }
    
    [cell.textLabel setFont:[UIFont fontWithName:@"Helvetica" size:16]];
    [cell.textLabel setTextColor:[UIColor colorWithRed:0.4f green:0.4f blue:0.4f alpha:1.0f]];

    return(cell);
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    NSInteger fileIdx = [self.currentDirectory isEqualToString:self.baseDirectory] ? indexPath.row : indexPath.row - 1;
    [self dismissKeyboard];
    
    //Special case for ..
    if(![self.currentDirectory isEqualToString:self.baseDirectory] && indexPath.row == 0) {
        self.currentDirectory = [self.currentDirectory stringByDeletingLastPathComponent];
        [self updateFileList];
        return;
    }
    
    //If it is a directory, change to that directory
    if(self.fileList[fileIdx][NSFileType] == NSFileTypeDirectory) {
        self.currentDirectory = [self.currentDirectory stringByAppendingPathComponent:self.fileList[fileIdx][@"FILENAME"]];
        [self updateFileList];
        self.selectedFilename = @"";
    } else { //Fill in the name in the filename field
        [self.inputTextField setText:self.fileList[fileIdx][@"FILENAME"]];
        self.selectedFilename = [self.currentDirectory stringByAppendingPathComponent:self.fileList[fileIdx][@"FILENAME"]];
    }
    
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

#pragma mark - UITextField delegate
- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    //Special case for ..
    if(![self.currentDirectory isEqualToString:self.baseDirectory] && [textField.text isEqualToString:@".."]) {
        self.currentDirectory = [self.currentDirectory stringByDeletingLastPathComponent];
        self.selectedFilename = @"";
        textField.text = @"";
        [self updateFileList];
        return(YES);
    }
    
    for(int i = 0; i < self.fileList.count; i++) {
        NSDictionary *file = self.fileList[i];
        if([file[@"FILENAME"] compare:textField.text options:NSLiteralSearch] == NSOrderedSame) {
            if(file[NSFileType] == NSFileTypeDirectory) {
                self.currentDirectory = [self.currentDirectory stringByAppendingPathComponent:textField.text];
                self.selectedFilename = @"";
                textField.text = @"";
                [self updateFileList];
            } else {
                [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForItem:i inSection:0] animated:YES scrollPosition:UITableViewScrollPositionTop];
                self.selectedFilename = [self.currentDirectory stringByAppendingPathComponent:textField.text];
                [self performSegueWithIdentifier:@"UnwindFileOpen" sender:nil];
            }
            return(YES);
        }
    }
    
    //File not found. Throw warning and clear entry if open dialog otherwise set values
    if(self.isOpenDialog) {
        UIAlertView *av = [[UIAlertView alloc] initWithTitle:@"Error" message:[NSString stringWithFormat:@"File not found:%@", textField.text] delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
        [av show];
        self.selectedFilename = @"";
        textField.text = @"";
    } else {
        self.selectedFilename = [self.currentDirectory stringByAppendingPathComponent:textField.text];
        [self performSegueWithIdentifier:@"UnwindFileOpen" sender:nil];
    }
    
    return(YES);
}

- (BOOL)disablesAutomaticKeyboardDismissal {
    return(NO);
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    NSString *textPath  = [self.currentDirectory stringByAppendingPathComponent:self.inputTextField.text];
    
    if(self.selectedFilename == nil || [self.selectedFilename compare:textPath options:NSLiteralSearch] != NSOrderedSame) {
        self.selectedFilename = textPath;
    }
}
@end

File: Samples/Swift/SignatureSDK-S/SignatureSDK-S/FileBrowser.swift
//
//  FileBrowser.swift
//  SignatureSDK-S
//
//  Created by Joss Giffard-Burley on 27/08/2014.
//  Copyright (c) 2014 Wacom. All rights reserved.
//

import Foundation

enum FSKeys {
    case FILENAME
    case PAT
}

class FileBrowser: UIViewController, UITableViewDataSource, UITableViewDelegate, UITextFieldDelegate {
    //==================================================================================================================
    // MARK: Class vars
    //==================================================================================================================
    
    @IBOutlet var cancelButton :UIButton!
    @IBOutlet var okButton :UIButton!
    @IBOutlet var tableView :UITableView!
    @IBOutlet var titleLabel :UILabel!
    @IBOutlet var inputTextField :UITextField!
    @IBOutlet var textFieldUnderline: UIView!
    @IBOutlet var logo: UIImageView!
    
    var selectedFilename :URL?
    var baseDirectory :URL?
    var inputFieldText :String?
    var titleText :String?
    var sourceFile :URL? //used for Open In... requests
    var isOpenDialog :Bool
    var signatureTypesOnly :Bool //Only display .txt, .fss & .png
    
    fileprivate let signatureExts = [ "fss", "png", "txt" ]
    
    fileprivate var directoryIcon = UIImage(named: "action_collection")?.tintedImageWithColour(UIColor(red: 0.121, green: 0.494, blue: 0.807, alpha: 1.0))
    fileprivate var binaryIcon = UIImage(named: "binary_file")?.tintedImageWithColour(UIColor(red: 0.121, green: 0.494, blue: 0.807, alpha: 1.0))
    fileprivate var textIcon = UIImage(named: "text_file")?.tintedImageWithColour(UIColor(red: 0.121, green: 0.494, blue: 0.807, alpha: 1.0))
    fileprivate var imageIcon = UIImage(named: "action_picture")?.tintedImageWithColour(UIColor(red: 0.121, green: 0.494, blue: 0.807, alpha: 1.0))
    fileprivate var currentDirectory :URL?
    
    fileprivate var fileList :[([FileAttributeKey: Any], filename: String, ext: String)] = []
    
    //==================================================================================================================
    // MARK: Init methods
    //==================================================================================================================
    
    required init?(coder aDecoder: NSCoder) {
        isOpenDialog = true
        signatureTypesOnly = false
        super.init(coder: aDecoder)
    }
    
    //==================================================================================================================
    // MARK: UIViewController methods
    //==================================================================================================================
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        okButton.layer.cornerRadius = 5.0
        cancelButton.layer.cornerRadius = 5.0
        tableView.layer.cornerRadius = 3.0
        tableView.layer.borderColor = UIColor(red: 0.7, green: 0.7, blue: 0.7, alpha: 1.0).cgColor
        tableView.layer.borderWidth = 1.0
        logo.image = UIImage(named: "logo_window")?.tintedImageWithColour(UIColor(red: 0.4, green: 0.4, blue: 0.4, alpha: 1.0))
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        titleLabel.text = titleText
        inputTextField.text = inputFieldText
        currentDirectory = baseDirectory
        tableView.delegate = self
        tableView.dataSource = self
        updateFileList()
    }
    
    //==================================================================================================================
    // MARK: Instance Methods
    //==================================================================================================================
    
    func updateFileList() {
        let fm = FileManager.default
        if let dirEnum = fm.enumerator(atPath: currentDirectory!.path) {
            
            fileList.removeAll(keepingCapacity: false)
            
            while let f :AnyObject = dirEnum.nextObject() as AnyObject? {
                let currentFile :URL = currentDirectory!.appendingPathComponent(f as! String)
                
                //Skip files starting '.' and subpaths
                if f.hasPrefix(".") || f.components(separatedBy: "/").count > 1 {
                    continue
                }
                
                if signatureTypesOnly {
                    if (signatureExts.contains(currentFile.pathExtension.lowercased()) || dirEnum.fileAttributes![FileAttributeKey.type] as? FileAttributeType == FileAttributeType.typeDirectory) {
                        let file = dirEnum.fileAttributes!
                        fileList.append((file, currentFile.lastPathComponent, currentFile.lastPathComponent))
                    }
                } else {
                    let file = dirEnum.fileAttributes!
                    fileList.append((file,  currentFile.pathExtension, currentFile.lastPathComponent ))
                }
                
            }
            
            //Sort file list A-Z
            fileList.sort(by: { (a, b) -> Bool in
                let one = a.filename
                let two = b.filename
                return one.compare(two, options:.caseInsensitive, range: nil, locale: nil) == ComparisonResult.orderedAscending
            })
            
            tableView.reloadData()
        }
    }
    
    func dismissKeyboard() {
        inputTextField.resignFirstResponder()
    }
    
    //==================================================================================================================
    // MARK: UITableViewDataSource delegate methods
    //==================================================================================================================
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "FILEBROWSER", for: indexPath) 
        let idx = baseDirectory == currentDirectory ? indexPath.row : indexPath.row - 1 //Add .. entry at the top of the list
        let imageExts = ["jpg", "png", "bmp", "gif", "jpg", "jpeg", "tiff"]
        let docExts = ["txt", "doc", "pdf", "docx", "pages"]
     
        
        if(indexPath.row == 0 && currentDirectory != baseDirectory) {
            cell.textLabel!.text = ".."
            cell.accessoryType = .none
            cell.imageView!.image = directoryIcon
        } else {
            var fileDetails = fileList[idx]
           
            if fileDetails.0[FileAttributeKey.type] as? FileAttributeType == FileAttributeType.typeDirectory { //Directory
                cell.imageView!.image = directoryIcon
                cell.accessoryType = .disclosureIndicator
            } else if imageExts.contains(fileDetails.ext) { //Image
                cell.imageView!.image = imageIcon
                cell.accessoryType = .none
            } else if docExts.contains(fileDetails.ext) { //Doc
                cell.imageView!.image = textIcon
                cell.accessoryType = .none
            } else {
                cell.imageView!.image = binaryIcon
                cell.accessoryType = .none
            }
            
            cell.textLabel!.text = fileDetails.filename
        }
        
        cell.textLabel!.font = UIFont(name: "Helvetica", size: 16)
        cell.textLabel!.textColor = UIColor(red: 0.4, green: 0.4, blue: 0.4, alpha: 1.0)
        
        return cell
    }
    
    //==================================================================================================================
    // MARK: UITableView delegate methods
    //==================================================================================================================
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        dismissKeyboard()
        
        //Special case for ..
        if (currentDirectory != baseDirectory) && indexPath.row == 0 {
            currentDirectory = currentDirectory?.deletingLastPathComponent()
            updateFileList()
            return
        }
        
        let idx = baseDirectory == currentDirectory ? indexPath.row : indexPath.row - 1 //Add .. entry at the top of the list
        let filename = fileList[idx].filename
        
        //If it is a directory, change to that directory
        if fileList[idx].0[FileAttributeKey.type]! as? FileAttributeType == FileAttributeType.typeDirectory {
            currentDirectory = currentDirectory?.appendingPathComponent(filename)
            updateFileList()
        } else { //Fill in the filename field
            inputTextField.text = filename
            selectedFilename = currentDirectory?.appendingPathComponent(filename)
        }
        
        tableView.deselectRow(at: indexPath, animated: true)
    }
    
    func tableView(_ tableView: UITableView, viewForFooterInSection section: Int) -> UIView? {
        return UIView()
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if baseDirectory == currentDirectory {
            return fileList.count
        } else {
            return fileList.count + 1
        }
    }
    
    //==================================================================================================================
    // MARK: UITextFieldDelegate methods
    //==================================================================================================================
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        //Special case for ..
        if currentDirectory != baseDirectory && textField.text == ".." {
            currentDirectory = currentDirectory?.deletingLastPathComponent()
            selectedFilename = nil
            textField.text = nil
            updateFileList()
            return true
        }
        
        var i = 0
        
        for file in fileList {
            if file.filename == textField.text {

                if file.0[FileAttributeKey.type]! as? FileAttributeType == .typeDirectory {
                    currentDirectory = currentDirectory?.appendingPathComponent(textField.text!)
                    selectedFilename = nil
                    textField.text = nil
                    updateFileList()
                } else {
                    tableView.selectRow(at: IndexPath(row: i, section: 0), animated: true, scrollPosition: .top)
                    selectedFilename = currentDirectory?.appendingPathComponent(textField.text!)
                    performSegue(withIdentifier: "UnwindFileOpen", sender: nil)
                }
                return true
            }
            i += 1
        }
        
        //File not found, throw warning and clear entry if open dialog
        if isOpenDialog {
            let av = UIAlertController(title: "Error", message: "File not found:\(String(describing: textField.text))", preferredStyle: .alert)
            av.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
            dismiss(animated: false, completion: nil)
            present(av, animated: true, completion: nil)
            updateFileList()
        } else {
            selectedFilename = currentDirectory?.appendingPathComponent(textField.text!)
            performSegue(withIdentifier: "UnwindFileOpen", sender: nil)
        }
        
        return true
    }
    
    //==================================================================================================================
    // MARK: Navigation methods
    //==================================================================================================================
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any!) {
        let textPath = currentDirectory?.appendingPathComponent(inputTextField.text!)
        
        if(selectedFilename == nil || selectedFilename != textPath) {
            selectedFilename = textPath
        }
    }
}
